### PR TITLE
[TechDocs] Initial Addon Framework Implementation

### PR DESCRIPTION
**Pull request by iamEAP**
## Hey, I just made a Pull Request!

Initial implementation of the TechDocs addon framework:

- Leverages the core extension API to encapsulate addon logic and display as react components
- Adds a `<TechDocsAddons>` registry component for composing/configuring addons.
- Exposes an alternative, addon-aware reader implementation

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/techdocs-addons/src/reader.tsx [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+      </Content>
+    );
+  }
+
+  return (
+    <Content>
+      {/* sheetsManager={new Map()} is needed in order to deduplicate the injection of CSS in the page. */}
+      <StylesProvider jss={jss} sheetsManager={new Map()}>
+        <div ref={ref} data-testid="techdocs-native-shadowroot" />
+        <Portal container={primarySidebarAddonSpace}>
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/reader.tsx:175]:
```suggestion
        <Portal container={primarySidebarAddonLocation}>
```

File: plugins/techdocs-addons/src/reader.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+          {addons.renderComponentsWithLocation(locations.PRIMARY_SIDEBAR)}
+        </Portal>
+        <Portal container={contentElement}>
+          {addons.renderComponentsWithLocation(locations.CONTENT)}
+        </Portal>
+        <Portal container={secondarySidebarAddonSpace}>
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/reader.tsx:181]:
```suggestion
        <Portal container={secondarySidebarAddonLocation}>
```

File: plugins/techdocs-addons/README.md [hunk 3]
```diff
@@ -1,5 +1,61 @@
...
+## Create a new addon
+
+To create a new addon, you can use the `createTechDocsAddon` factory exported
+from this plugin. Normally, addons are provided by Backstage plugins, which can
+then be composed within a Backstage app.
+
+When you create a new Addon, it requires three things.
+
+1. A `name` for debugging and analytics purposes)
+2. A `location`, indicating where/how the addon will be rendered
```
Comment (emmaindal) [inline plugins/techdocs-addons/README.md:21]: maybe we can add the available locations here too

File: plugins/techdocs-addons/src/addons.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  const findAddonByData = useCallback(
+    (data: TechDocsAddonOptions | undefined) => {
+      if (!collection || !data) return null;
+      const nameKey = getDataKeyByName(data.name);
+      return getTechDocsAddonByName(collection, nameKey) ?? null;
+    },
+    [collection],
+  );
+
+  const renderComponentWithName = useCallback(
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/addons.tsx:103]: sorry for nitpicking but now when I think about it maybe
renderComponentWithName --> renderComponentByName?

File: plugins/techdocs-addons/src/addons.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    (name: string) => {
+      const data = options.find(option => option.name === name);
+      return data ? findAddonByData(data) : null;
+    },
+    [options, findAddonByData],
+  );
+
+  const renderComponentsWithLocation = useCallback(
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/addons.tsx:111]: same as above

File: plugins/techdocs-addons/src/context.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+  children,
+}: PropsWithEntityName) => {
+  const { NotFoundErrorPage } = useApp().getComponents();
+  const techdocsApi = useApi(techdocsApiRef);
+
+  const { value, loading, error } = useAsync(async () => {
+    return await techdocsApi.getTechDocsMetadata(entityName);
+  }, []);
+
+  if (!loading && error) {
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/context.tsx:51]: what do you think about making the loading and error states available through the context instead and let the ones consuming it be responsible for error handling?

File: plugins/techdocs-addons/src/context.tsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  children,
+}: PropsWithEntityName) => {
+  const { NotFoundErrorPage } = useApp().getComponents();
+  const techdocsApi = useApi(techdocsApiRef);
+
+  const { value, loading, error } = useAsync(async () => {
+    return await techdocsApi.getEntityMetadata(entityName);
+  }, []);
+
+  if (!loading && error) {
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/context.tsx:86]: what do you think about making the loading and error states available through the context instead and let the ones consuming it be responsible for error handling?

File: plugins/techdocs-addons/src/reader.tsx [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+import {
+  TechDocsMetadataProvider,
+  useMetadata,
+  TechDocsEntityProvider,
+  TechDocsReaderPageProvider,
+  useTechDocsReaderPage,
+} from './context';
+import { TechDocsAddonLocations as locations } from './types';
+
+const TechDocsReaderPageSubheader = withStyles(theme => ({
```
Comment (emmaindal) [inline plugins/techdocs-addons/src/reader.tsx:52]: what do you think about separate these components into its own files? for improved readability?

**Review by emmaindal — COMMENTED** [review]
🎉

**Comment by iamEAP** [comment]
Thanks for the review @emmaindal!  Reorganization + smaller feedback handled here ([573a05d](https://github.com/backstage/backstage/pull/10203/commits/573a05dce505fe213903bce56155caaa142f82fd)) and then this commit exposes the error/loading status of metadata hooks to callers ([7c99d76](https://github.com/backstage/backstage/pull/10203/commits/7c99d76fa5a3670849d4b1dda956e1f27fea57b3)).

**Pull request merged**